ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```
@@ -0,0 +1,350 @@
+    if (!compressionConfig.isCompressionEnabled) {
+      return false;
+    }
+
+    // Check content-encoding. BlobProperty.ContentEncoding contains a string, do not apply compression.
+    // Note: Calling getBlobProperties() throws if blob properties is null, so it never returns null.
+    String contentEncoding = blobProperties.getContentEncoding();
+    if (compressionConfig.isSkipWithContentEncoding && !Utils.isNullOrEmpty(contentEncoding)) {
+      logger.info("No compression applied because compression content is encoded with " + contentEncoding);
+      // Emit metrics to count how many compressions skipped due to encoded data.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipContentEncoding.inc();
+      return false;
+    }
+
+    // Check the content-type.
+    String contentType = blobProperties.getContentType();
+    if (!compressionConfig.isCompressibleContentType(contentType)) {
+      logger.info("No compression applied because content-type " + contentType + " is incompressible.");
+      // Emit metrics to count how many compressions skipped due to incompressible chunks based on content-type.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipContentTypeFiltering.inc();
+      return false;
+    }
+
+    // For all other case, compression should be enabled.
+    return true;
+  }
+
+  /**
+   * Apply compression to a chunk buffer if compression is enabled and content-type is compressible.
+   * It applies compression config to determine whether to accept or reject the compressed result.
+   * For example, if the compressed data size is smaller than original data by certain % (exceeds threshold),
+   * the compressed data is accepted and return compressed data; otherwise, compression is discarded.
+   * This method emits metrics along the process.  It returns null if failed instead of throwing exceptions.
```

I like that it returns null instead of throwing exc

---

I agree.  I would throw an exception if something went wrong and the code can't handle it.  I would return null if it's expected and can be handled by the method.